controller/hybridgateway/target/target.go

```diff
@@ -0,0 +1,411 @@
+		return []configurationv1alpha1.KongTarget{}, nil
+	}
+
+	// Step 2: Recalculate weights across all valid BackendRefs.
+	validBackendRefs = recalculateWeightsAcrossBackendRefs(validBackendRefs)
+
+	// Step 3: Create KongTargets from the processed ValidBackendRef structs.
+	targets, err := createTargetsFromValidBackendRefs(httpRoute, pRef, upstreamName, validBackendRefs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create targets from valid BackendRefs: %w", err)
+	}
+
+	logger.V(1).Info("created targets for BackendRefs",
+		"totalBackendRefs", len(backendRefs),
+		"validBackendRefs", len(validBackendRefs),
+		"createdTargets", len(targets))
+
+	return targets, nil
+}
+
+// findBackendRefPortInService returns the ServicePort from svc that matches the port specified in bRef.
+// If bRef.Port is nil or no matching port is found in svc.Spec.Ports, an error is returned.
+// This function is used to validate and resolve the actual ServicePort for a given BackendRef.
+func findBackendRefPortInService(bRef *gwtypes.HTTPBackendRef, svc *corev1.Service) (*corev1.ServicePort, error) {
+	// Check if the port is specified in the BackendRef. The port is required.
+	if bRef.Port == nil {
+		// If the port is not specified, return an error.
+		return nil, fmt.Errorf("port not specified in BackendRef")
+	}
+
+	// Find the port in the service that matches the port in the BackendRef.
+	svcPort, svcPortFound := lo.Find(svc.Spec.Ports, func(p corev1.ServicePort) bool {
+		return p.Port == int32(*bRef.Port)
+	})
+	if !svcPortFound {
+		// If the port is not found, return an error.
+		return nil, fmt.Errorf("port %v not found in service %s/%s", *bRef.Port, svc.Namespace, svc.Name)
+	}
+
+	return &svcPort, nil
+}
+
+// getEndpointSlicesForService retrieves all EndpointSlices for a given service.
+func getEndpointSlicesForService(ctx context.Context, cl client.Client, svc *corev1.Service) (*discoveryv1.EndpointSliceList, error) {
+	endpointSlices := &discoveryv1.EndpointSliceList{}
+	req, err := labels.NewRequirement(discoveryv1.LabelServiceName, selection.Equals, []string{svc.Name})
+	if err != nil {
+		return nil, err
+	}
+	labelSelector := labels.NewSelector().Add(*req)
+	err = cl.List(ctx, endpointSlices, &client.ListOptions{Namespace: svc.Namespace, LabelSelector: labelSelector})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list endpointslices for service %s/%s: %w", svc.Namespace, svc.Name, err)
+	}
+	return endpointSlices, nil
+}
+
+// extractReadyEndpointAddresses extracts all ready endpoint addresses that match the service port.
+func extractReadyEndpointAddresses(endpointSlices *discoveryv1.EndpointSliceList, svcPort *corev1.ServicePort) []string {
+	var addresses []string
+
+	for _, endpointSlice := range endpointSlices.Items {
+		for _, p := range endpointSlice.Ports {
+			// Skip ports that don't match the service port.
+			if p.Port == nil || *p.Port < 0 || *p.Protocol != svcPort.Protocol || *p.Name != svcPort.Name {
```

Comment:
why `Port<0`? Shouldn't we check `p.port == svcPort.Port`?

Reply:
This case suits the headless services without selectors: https://github.com/Kong/kong-operator/issues/2516